File: leetcode/binary_search/3155_Maximum_Number_of_Upgradable_Servers/codes/bs.py
from typing import List

class Solution:
    def maxUpgrades(self, count: List[int], upgrade: List[int], sell: List[int], money: List[int]) -> List[int]:
        def can(mid, c, u, s, m):
            need_sell = c - mid
            total = m + need_sell * s
            return mid * u <= total

        n = len(count)
        result = []
        for i in range(n):
            low, high = 0, count[i]
            best = 0
            while low <= high:
                mid = (low + high) // 2
                if can(mid, count[i], upgrade[i], sell[i], money[i]):
                    best = mid
                    low = mid + 1
                else:
                    high = mid - 1
            result.append(best)
        return result

File: leetcode/binary_search/3155_Maximum_Number_of_Upgradable_Servers/codes/bs.c
#include <stdlib.h>

int* maxUpgrades(int* count, int countSize, int* upgrade, int upgradeSize, int* sell, int sellSize, int* money, int moneySize, int* returnSize) {
    int* result = (int*)malloc(sizeof(int) * countSize);

    for (int i = 0; i < countSize; ++i) {
        int low = 0, high = count[i], best = 0;
        while (low <= high) {
            int mid = (low + high) / 2;
            int needSell = count[i] - mid;
            long long total = (long long)money[i] + (long long)needSell * sell[i];
            if ((long long)mid * upgrade[i] <= total) {
                best = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        result[i] = best;
    }

    *returnSize = countSize;
    return result;
}
